dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -204,31 +210,79 @@ void Microsoft::Windows::AppNotifications::Helpers::RegisterAssets(std::wstring
         &hKey,
         nullptr /* lpdwDisposition */));
 
+    // Retrieve DisplayName and IconUri
+    // 1. We call into Shell APIs, to get both from the app taskbar shortcut.
+    // 2. If 1 doesn't work (i.e. possibly because the developer didn't specify any DisplayName/Icon,
+    //    we take the file name as app name, and let Shell to put a default icon
+    winrt::com_ptr<IPropertyStore> propertyStore;
+    wil::unique_hwnd hWindow{ GetConsoleWindow() };
+    THROW_IF_FAILED(SHGetPropertyStoreForWindow(hWindow.get(), IID_PPV_ARGS(propertyStore.put())));
+
+    // Retrieve the display name
+    std::wstring displayName{};
+
+    wil::unique_prop_variant propVariantDisplayName;
+    THROW_IF_FAILED(propertyStore->GetValue(PKEY_AppUserModel_RelaunchDisplayNameResource, &propVariantDisplayName));
+
+    displayName = propVariantDisplayName.pwszVal != nullptr ? propVariantDisplayName.pwszVal : L"";
+
+    if (displayName.length() == 0)
+    {
+        wil::unique_cotaskmem_string processName;
+        THROW_IF_FAILED(GetCurrentProcessPath(processName));
+
+        displayName = processName.get();
+
+        size_t lastBackslashPosition{ displayName.rfind(L"\\") };
+        THROW_HR_IF(E_UNEXPECTED, lastBackslashPosition == std::wstring::npos);
+        displayName = displayName.substr(lastBackslashPosition + 1); // One after the delimiter
+
+        displayName.erase(displayName.find_first_of(L".")); // Remove file extension
+    }
+
+    // Retrieve the icon
+    std::wstring iconFilePath{};
+
+    wil::unique_prop_variant propVariantIcon;
+    THROW_IF_FAILED(propertyStore->GetValue(PKEY_AppUserModel_RelaunchIconResource, &propVariantIcon));
+    iconFilePath = propVariantIcon.pwszVal != nullptr ? propVariantIcon.pwszVal : L"";
+
+    // Icon filepaths from Shell APIs have this format: <filepath>,-<index>,
+    // since .ico files can have multiple icons in the same file.
+    // NotificationController doesn't seem to support such format, so let it take the first icon by default.
+    auto iterator = iconFilePath.find_first_of(L",");
+    if (iterator != std::wstring::npos)
```

Comment:
THROWHR_IF(E_UNEXPECTED, iterator == std::wstring::npos)

Reply:
Not necessarily. If the dev provides a plain png filepath without an index, it is still a valid case.